### PR TITLE
Freebase Entity Index

### PR DESCRIPTION
**Pull request by salman1993**
- builds up the index for freebase entities
- indexes three fields: 1) wikipedia_en_title, 2) rdf#label, 3) wikipedia_en (may contain variants)
- search class to look up an entity ID given a query

### CODE REVIEW

**Comment by lintool** [comment]
You inadvertently checked-in a whole bunch of things that should be there, e.g., `freebase.sample.index/`, `freebase-subject-obama.gz`

File: src/main/java/io/anserini/index/generator/LuceneFreebaseEntityDocumentGenerator.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    // Index subject as a StringField to allow searching
+    Field entityField = new StringField(FIELD_ENTITY, cleanUri(tripleDoc.getEntityId()), Field.Store.YES);
+    doc.add(entityField);
+
+    FieldType fieldType = new FieldType();
+    fieldType.setStored(true);
+    fieldType.setStoreTermVectors(true);
+    fieldType.setIndexOptions(IndexOptions.DOCS_AND_FREQS);
+
+    Field titleField = new TextField(FIELD_TITLE, tripleDoc.getTitle(), Field.Store.YES);
```
Comment (lintool) [inline src/main/java/io/anserini/index/generator/LuceneFreebaseEntityDocumentGenerator.java]: What's the difference between the `Field_TITLE` and `FIELD_LABEL`?

File: src/main/java/io/anserini/search/SearchFreebaseEntity.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import org.kohsuke.args4j.*;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+
+import static io.anserini.index.generator.LuceneDocumentGenerator.FIELD_BODY;
```
Comment (lintool) [inline src/main/java/io/anserini/search/SearchFreebaseEntity.java]: I don't think you need these?

File: src/main/java/io/anserini/index/generator/LuceneFreebaseEntityDocumentGenerator.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import io.anserini.index.IndexFreebaseEntityCollection;
+import org.apache.lucene.document.*;
+import org.apache.lucene.index.IndexOptions;
+
+/**
+ * Converts a {@link RDFDocument} into
+ * a Lucene {@link Document}.
+ */
+public class LuceneFreebaseEntityDocumentGenerator {
+  public static final String FIELD_ENTITY = "entity";
```
Comment (lintool) [inline src/main/java/io/anserini/index/generator/LuceneFreebaseEntityDocumentGenerator.java]: document what each of these fields are?

**Comment by lintool** [comment]
Additional comment - I think we should use the Freebase terminology more precisely. I believe "entities" are called "topics" - so this should be called "FreebaseTopicIndex". We should also use `mid`, etc.

File: src/main/java/io/anserini/index/generator/LuceneFreebaseEntityDocumentGenerator.java (hunk 1 above)
Comment (salman1993) [inline src/main/java/io/anserini/index/generator/LuceneFreebaseEntityDocumentGenerator.java]: I'll put in comments and in the docs

**Comment by salman1993** [comment]
@minafarid - yeah i'm sure about the dots format. This is what it says in https://developers.google.com/freebase/:
"Note: In Freebase, objects have MIDs that look like /m/012rkqx. In RDF those MIDs become m.012rkqx."

**Comment by minafarid** [comment]
@salman1993 my problem was in the prefix of the URI, not the dot notation. But it is okay if the data is represented this way.

**Pull request closed without merging**